src/proof.rs

```diff
@@ -0,0 +1,277 @@
+// Copyright 2020 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::nibbleslice::NibbleSlice;
+use crate::node::Node;
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use primitives::Bytes;
+use primitives::H256;
+
+// Unit of a proof.
+//#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofUnit {
+    pub root: H256,
+    pub key: H256,
+    pub value: Option<Bytes>, // None in case of absence
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProof(pub Vec<Bytes>);
+
+pub trait CryptoStructure {
+    fn make_proof(&self, key: &H256) -> crate::Result<(CryptoProofUnit, CryptoProof)>;
+}
+
+/// A verification logic of TrieDB's Merkle proof.
+/// For the format of proof, check the make_proof() function.
+/// It verifies the proof with a given unit of test.
+/// It should never abort or fail, but only return 'false' as result of getting invalid or ill-formed proof.
```

Comment:
It should never abort or fail, but only return 'false' as a result of getting an invalid or ill-formed proof.